### PR TITLE
[SchedulerDesign] New scheduler design

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Alternative to

### CODE REVIEW

File: src/diffusers/schedulers/scheduling_lms_discrete.py [hunk 1]
```diff
@@ -178,7 +179,8 @@ def step(
             When returning a tuple, the first element is the sample tensor.
 
         """
-        sigma = self.sigmas[timestep]
+        index = (self.config.num_train_timesteps - timestep) // (self.config.num_train_timesteps // self.num_inference_steps)
```
Comment (keturn) [inline src/diffusers/schedulers/scheduling_lms_discrete.py:182]: 👍 I like this bit, having the scheduler responsible for figuring out what to do with the timestep instead of having the pipeline keep track of how schedulers interpret their arguments.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -259,8 +259,8 @@ def __call__(
             timesteps_tensor = torch.tensor(self.scheduler.timesteps.copy(), device=self.device)
 
         # if we use LMSDiscreteScheduler, let's make sure latents are multiplied by sigmas
-        if isinstance(self.scheduler, LMSDiscreteScheduler):
-            latents = latents * self.scheduler.sigmas[0]
+        if self.scheduler.type == SchedulerType.CONTINUOUS:
+            latents = latents * self.scheduler.init_sigma
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:263]: A step in the right direction. I'd love to get rid of the `if` entirely, but as long as we have it, defining a scheduler.type enum is much preferable to isinstance!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -274,10 +274,9 @@ def __call__(
         for i, t in enumerate(self.progress_bar(timesteps_tensor)):
             # expand the latents if we are doing classifier free guidance
             latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
-            if isinstance(self.scheduler, LMSDiscreteScheduler):
-                sigma = self.scheduler.sigmas[i]
-                # the model input needs to be scaled to match the continuous ODE formulation in K-LMS
-                latent_model_input = latent_model_input / ((sigma**2 + 1) ** 0.5)
+
+            if self.scheduler.type == SchedulerType.CONTINUOUS and self.model.config.trained_scheduler_type == SchedulerType.DISCRETE:
+                latent_model_input = self.scheduler.scale_(latent_model_input)
```
Comment (keturn) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:279]: is this `scale_` implemented?

**Review by anton-l — COMMENTED** [review]
@patrickvonplaten as you said in #637:
> I think we should focus much more on global differences

Which isn't solved fully here for a couple of reasons (that I see at the moment):
* Putting the schedulers into continuous and discrete categories is a bit restrictive IMO, as it implies that these categories have different APIs, while mathematically there's no such restriction (we can freely convert between the timestep representations, as Karras et al. have shown). Having this distinction on the model side can be beneficial to learn what type of noise conditioning the model expects, but we can get away with having it on the pipeline side too.
* With this design we still have a very different representation of `timesteps` between different schedulers: sometimes they're int, sometimes they're float, sometimes they're descending, sometimes ascending. #637 addresses this by allowing the schedulers to have their internal notion of `schedule` (or `sigmas, or whatever they choose), while only having integer descending `timesteps` in the public API. This makes debugging the pipelines way easier (i.e. timesteps no longer jump all over the place, we only have indices/steps), while also ensuring that literally any scheduler (with further refactoring for our VP, VE and Karras) is usable with e.g. Stable Diffusion.

**Comment by patrickvonplaten** [comment]
@anton-l :

RE:

> Putting the schedulers into continuous and discrete categories is a bit restrictive IMO

-> I don't understand this. All of our schedulers belong always to one of the two classes no? The math doesn't have to be perfect there it's just important that people understand what timestep representation is expected. We can also find a better name

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -288,10 +287,7 @@ def __call__(
                 noise_pred = noise_pred_uncond + guidance_scale * (noise_pred_text - noise_pred_uncond)
 
             # compute the previous noisy sample x_t -> x_t-1
-            if isinstance(self.scheduler, LMSDiscreteScheduler):
-                latents = self.scheduler.step(noise_pred, i, latents, **extra_step_kwargs).prev_sample
-            else:
-                latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
+            latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs).prev_sample
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:290]: Could be considered a bug correction IMO

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -274,10 +274,9 @@ def __call__(
         for i, t in enumerate(self.progress_bar(timesteps_tensor)):
             # expand the latents if we are doing classifier free guidance
             latent_model_input = torch.cat([latents] * 2) if do_classifier_free_guidance else latents
-            if isinstance(self.scheduler, LMSDiscreteScheduler):
-                sigma = self.scheduler.sigmas[i]
-                # the model input needs to be scaled to match the continuous ODE formulation in K-LMS
-                latent_model_input = latent_model_input / ((sigma**2 + 1) ** 0.5)
+
+            if self.scheduler.type == SchedulerType.CONTINUOUS and self.model.config.trained_scheduler_type == SchedulerType.DISCRETE:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:278]:
```suggestion
            if self.scheduler.type == SchedulerType.CONTINUOUS and self.model.config.trained_scheduler_type == SchedulerType.DISCRETE:
```

Here I'm very open to hear better suggestions @anton-l (if you can think about something that doesn't require changes to DDIM or DDPM but that's cleaner)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:263]: @anton-l can also be convinced here to change it to something that might be cleaner (again if we don't have to force a function upon DDIM or DDPM)

Overall, I feel quite strongly about the following though:
- Let's not make easy schedulers more complex because we'd like to support newer, more complex schedulers
- Forcing every scheduler to implement a certain method that can grow arbitrary in complexity is much worse that **educating** if statements with a nice comment.

**Comment by keturn** [comment]
There's a _lot_ of back-and-forth about that _timestamps_ list and what its internal representation is and what the interfaces to it look like—

but is it necessary for a Scheduler to expose that list at all?  
What benefits does that provide?

There might be some. I'm not sure the current pipelines need them at all. As far as I can tell, the only thing they need to know is the _t_ of the current step so they can pass it to `model.forward`.

It also embodies the assumption that there _is_ a fixed schedule known up front, which doesn't need to be the case. For example https://github.com/LuChengTHU/dpm-solver implements a sampler that uses adaptive time steps— it just keeps going until it's _close enough_ to 0.

That's the sort of thing I wasn't going to bring up in this early pass of the redesign, but I do so now because it seems like it might be more _freeing_ than _complicating._

**Comment by patrickvonplaten** [comment]
> There's a _lot_ of back-and-forth about that _timestamps_ list and what its internal representation is and what the interfaces to it look like—
> 
> but is it necessary for a Scheduler to expose that list at all? What benefits does that provide?
> 
> There might be some. I'm not sure the current pipelines need them at all. As far as I can tell, the only thing they need to know is the _t_ of the current step so they can pass it to `model.forward`.
> 
> It also embodies the assumption that there _is_ a fixed schedule known up front, which doesn't need to be the case. For example https://github.com/LuChengTHU/dpm-solver implements a sampler that uses adaptive time steps— it just keeps going until it's _close enough_ to 0.
> 
> That's the sort of thing I wasn't going to bring up in this early pass of the redesign, but I do so now because it seems like it might be more _freeing_ than _complicating._

Thanks for bringing it up! Regarding the current design I think we can assume though that timesteps can always be written as a list of values. Timesteps cannot be written as a list of values are exceptional for now IMO (these things can't be traced etc... so not sure how useful they are)

**Comment by patrickvonplaten** [comment]
Closing this PR in favor of: https://github.com/huggingface/diffusers/pull/719

**Comment by keturn** [comment]
>  (these things can't be traced etc... so not sure how useful they are)

What does _traced_mean in this context?

The only disadvantage I could think of to giving up a fixed number of iterations for something adaptive is if there was some JIT compiler doing some loop unrolling or parallelization or something. But CPython isn't that smart, and it's not parallelizable because each iteration depends on the output of the one previous.

**Comment by patrickvonplaten** [comment]
> > (these things can't be traced etc... so not sure how useful they are)
> 
> What does _traced_ mean in this context?
> 
> The only disadvantage I could think of to giving up a fixed number of iterations for something adaptive is if there was some JIT compiler doing some loop unrolling or parallelization or something. But CPython isn't that smart, and it's not parallelizable because each iteration depends on the output of the one previous.

Yeah sorry I more or less meant this by "traced" - JAX won't like this and also not sure if ONNX would be happy about this

**Comment by keturn** [comment]
Oh, do we have to play by [these JAX rules][jax-sharp]? Good to know.

…though I'm _skeptical_ about the value of JIT-ing that whole inference loop, and even if you did want to, `jax.lax.while_loop` exists…

But we can leave that until later. No need to worry over how to implement the adaptive form of DPM-Solver before we're even able to implement the more predictable DPM-Solver-fast sampler.

[jax-sharp]: https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html "🔪 JAX - The Sharp Bits 🔪"

**Comment by patrickvonplaten** [comment]
> implement

We also don't have to play by the JAX rules at all ;-) We want to support JAX as soon as it runs on free Google colabs: https://github.com/googlecolab/colabtools/issues/3009#issuecomment-1254950921 so that we give users more power (8 TPUs are pretty powerful). That being said it doesn't mean that all PyTorch functionality will have to take into account how the mirror would work in JAX. It's totally fine if the two frameworks diverge. 

More generally, it allows helps PyTorch optimization libraries like ONNX, TensorRT a lot though if all memory can be pre-allocated.
